### PR TITLE
#1082 Add cache check for external repositories used during testing

### LINKED ISSUE #1082: Cache build dependencies in CI
Sometimes, testing fails because it can't access an online resource. Caching them (using [`actions/cache`](https://github.com/actions/cache)) would avoid that problem, as well as speed up testing.

Candidates for caching:
- [ ] The external repos pulled by `test/run-tests.sh`
- [ ] The libpng sources pulled by `.github/actions/mingw-w64-libpng-dev.sh`

...any anything else that gets `wget`/`curl`/`git clone`d and that I missed.

### PR DESCRIPTION
**Pull request by rlewicki**
Hi, I've prepared a PR that checks for a cached external repositories. The approach is based on a comments you mentioned in #1082. I hope  this is something you had in mind, if not please do let me know how I can adjust the solution.

Perhaps you would prefer each external repository hash to be stored in a separate file so if only one of them changes we only need to update one repository cache? Not sure if it's worth going this granular?

Fixes #1082.

### CODE REVIEW

File: test/fetch-test-deps.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+set -e
+
+cd "$(dirname "$0")"
+
+echo "Fetching test dependency repositories"
+
+test_downstream() { # owner/repo shallow-since commit make-target
```
Comment (Rangi42) [inline test/fetch-test-deps.sh:9]:
```suggestion
fetch_downstream() { # owner/repo shallow-since commit
```

File: test/run-tests.sh [hunk 2]
```diff
@@ -26,19 +26,12 @@ done
 # to the day before, to reduce the amount of refs being transferred and thus speed up CI.
 
 test_downstream() { # owner/repo shallow-since commit make-target
```
Comment (Rangi42) [inline test/run-tests.sh:28]:
```suggestion
test_downstream() { # owner/repo make-target
```

**Comment by ISSOtm** [comment]
> Perhaps you would prefer each external repository hash to be stored in a separate file so if only one of them changes we only need to update one repository cache? Not sure if it's worth going this granular?

Considering how thus far, the repos have more or less always been updated in bulk, and the cloning operation is fairly fast, it's probably fine.

File: test/run-tests.sh [hunk 3]
```diff
@@ -26,19 +26,12 @@ done
...
-	if [ ! -d ${1##*/} ]; then
-		git clone https://github.com/$1.git --shallow-since=$2 --single-branch
-	fi
 	pushd ${1##*/}
```
Comment (ISSOtm) [inline test/run-tests.sh:29]:
```suggestion
	if ! pushd ${1##*/}; then
		echo >&2 'Please run `fetch-test-deps.sh` before running the test suite'
		return 1
	fi
```
Hinting the user at how to fix the problem should be better UX. Doing this on `pushd` fail is preferable, as they might have e.g. symlink'd the directories (I think).

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Thanks!

File: test/run-tests.sh (hunk 3 above)
Comment (rlewicki) [reply test/run-tests.sh:29]: Very good point, I haven't thought about giving an error message for users that will run the scripts manually. Thanks for the suggestion.

File: test/run-tests.sh [hunk 4]
```diff
@@ -25,20 +25,16 @@ done
...
+	make -j4 $2 RGBDS=../../
 	popd
 }
 
-test_downstream pret/pokecrystal 2022-09-29 70a3ec1accb6de1c1c273470af0ddfa2edc1b0a9 compare
-test_downstream pret/pokered     2022-09-29 2b52ceb718b55dce038db24d177715ae4281d065 compare
-test_downstream AntonioND/ucity  2022-04-20 d8878233da7a6569f09f87b144cb5bf140146a0f ''
+test_downstream pret/pokecrystal compare
+test_downstream pret/pokered compare
+test_downstream AntonioND/ucity ''
```
Comment (Rangi42) [inline test/run-tests.sh:40]: :pick: To match `fetch-test-deps.sh`'s alignment:
```suggestion
test_downstream pret/pokecrystal compare
test_downstream pret/pokered     compare
test_downstream AntonioND/ucity  ''
```

File: test/fetch-test-deps.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	git checkout -f $3
+	if [ -f ../patches/${1##*/}.patch ]; then
+		git apply --ignore-whitespace ../patches/${1##*/}.patch
+	fi
+	popd
+}
+
+fetch_downstream pret/pokecrystal 2022-09-29 70a3ec1accb6de1c1c273470af0ddfa2edc1b0a9
+fetch_downstream pret/pokered     2022-09-29 2b52ceb718b55dce038db24d177715ae4281d065
+fetch_downstream AntonioND/ucity  2022-04-20 d8878233da7a6569f09f87b144cb5bf140146a0f
```
Comment (ISSOtm) [inline test/fetch-test-deps.sh:23]:
```suggestion
fetch_downstream AntonioND/ucity  2022-04-20 d8878233da7a6569f09f87b144cb5bf140146a0f

```
Newline at EOF is preferable.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: .github/workflows/testing.yml [hunk 6]
```diff
@@ -139,6 +150,17 @@ jobs:
...
           name: rgbds-canary-win${{ matrix.bits }}
           path: bins
+      - name: Check test dependency repositories cache
+        id: test-deps-cache
+        uses: actions/cache@v3
+        with:
+          path: test
+          key: ${{ hashFiles('test/fetch-test-deps.sh') }}
+      - if: ${{ steps.test-deps-cache.outputs.cache-hit != 'true' }}
+        name: Fetch test dependency repositories
```
Comment (ISSOtm) [inline .github/workflows/testing.yml:160]:
```suggestion
        name: Fetch test dependency repositories
        shell: bash
```
I think this is why the Windows tests are failing.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: .github/workflows/testing.yml (hunk 6 above)
Comment (rlewicki) [reply .github/workflows/testing.yml:160]: Ah yes, probably. I ignored the failure error cause I wasn't up to date with my fork and it was failing tests due to not being able to download zlib, didn't notice it's actually failing on something else. My bad, sorry!

**Review by ISSOtm — APPROVED** [review]

**Review by Rangi42 — APPROVED** [review]

**Pull request merged**